netkan/netkan/indexer.py

```diff
@@ -154,45 +163,54 @@ def process_ckan(self) -> None:
         self._process_ckan()
 
     @property
-    def delete_attrs(self) -> Dict[str, Any]:
+    def delete_attrs(self) -> DeleteMessageBatchRequestEntryTypeDef:
         return {
             'Id': self.message_id,
             'ReceiptHandle': self.receipt_handle
         }
 
 
-class MessageHandler:
-
-    def __init__(self, repo: CkanMetaRepo, github_pr: Optional[GitHubPR] = None) -> None:
-        self.ckm_repo = repo
-        self.github_pr = github_pr
-        self.master: Deque[CkanMessage] = deque()
-        self.staged: Deque[CkanMessage] = deque()
-        self.processed: List[CkanMessage] = []
+class MessageHandler(BaseMessageHandler):
+    _master: Deque[CkanMessage]
+    _staged: Deque[CkanMessage]
+    _processed: List[CkanMessage]
 
     def __str__(self) -> str:
-        return str(self.master + self.staged)
+        return str(' '.join([str(x) for x in self.master + self.staged]))
 
     def __len__(self) -> int:
         return len(self.master + self.staged)
 
-    # Apparently gitpython can be leaky on long running processes
-    # we can ensure we call close on it and run our handler inside
-    # a context manager
-    def __enter__(self) -> 'MessageHandler':
-        if not self.ckm_repo.is_active_branch('master'):
-            self.ckm_repo.checkout_branch('master')
-        self.ckm_repo.pull_remote_branch('master')
-        return self
+    @property
+    def repo(self) -> CkanMetaRepo:
+        return self.game.ckanmeta_repo
 
-    def __exit__(self, exc_type: Type[BaseException],
-                 exc_value: BaseException, traceback: TracebackType) -> None:
-        self.ckm_repo.close_repo()
+    @property
+    def github_pr(self) -> GitHubPR:
+        return self.game.github_pr
 
-    def append(self, message: 'boto3.resources.factory.sqs.Message') -> None:
+    @property
+    def master(self) -> Deque[CkanMessage]:
+        if getattr(self, '_master', None) is None:
+            self._master = deque()
+        return self._master
+
+    @property
+    def staged(self) -> Deque[CkanMessage]:
+        if getattr(self, '_staged', None) is None:
+            self._staged = deque()
+        return self._staged
+
+    @property
+    def processed(self) -> List[CkanMessage]:
+        if getattr(self, '_processed', None) is None:
+            self._processed = []
+        return self._processed
+
```

Comment:
Are a `deque` or a list really large enough to be worth the added complexity of lazy initialization? This looks like textbook premature optimization.

Reply:
It's always been a deque. But it's not so much for optimisation, rather convenience. Allows the queue to be processed in the order it came in via `popleft`. It's worked for nearly 4 years, so I think I'm happy to keep it the way it is.

Reply:
Oh, I'm absolutely fine with using a `deque`. My question is whether these properties are needed, rather than just doing this in `__init__`:

```python
    self.master = deque()
    self.staged = deque()
    self.processed = []
```

What do we get from making these properties lazily initialized instead?

Reply:
It's less work to do within the init, and means we don't need to overwrite it during inheritance. It's wasn't so much for performance reasons, just a combination of my preference for lazy loading and a slightly cleaner implementation when it comes to inheritance.

Reply:
Inheritance would be handled thusly in the derived class:

```python
    def __init__(self) -> None:
        super().__init__(params)
```

(In most languages the compiler would do this for you, but Python is special. :snake:)

Reply:
I realise I didn't give a complete answer here. Mostly because I do a lot of this in the code I write at work. 

Type safety is the predominant reason, it means we can get rid of the Optional. The static definition at the top of the class only defines the type when you don't set it, hence the need to "lazy load".

It's my preference to be strict about what a property has, rather than needing to worry about it silently none (as far as the type checker is concerned).

Reply:
What I'm suggesting is perfectly type safe and doesn't require Optional. Sounds like a good reason not to use those static definitions.

Reply:
@HebaruSan oh, I confused this with something else I was looking at. Doing it this way avoids the need to override the `__init__` in the consuming class (due to the inherent type challanges, it does make it a consistent pattern with the challenge I mentioned). Which is generally preferable in Python land. The static analysis tools I use at work (paid) haven't ever marked this pattern as a code smell and pylint hasn't complained. There is nothing here that is inherently wrong beyond a difference in style preferences. I don't feel like it's worth refactoring further in this PR. If we're really motivated we can  circle back around with a better pattern if desired, as there are test cases.

Reply:
It's desired! Strongly desired!

Reply:
The alternative pattern to this would be to write custom descriptors or use a dataclass. Which for this use case seems *way* overkill to make this a little more "DRY", along with burying the logic further. I personally find a class more readable, and the added ability to inline docstrings when the properties aren't all buried up in the init. It is notable that the complaints from our static analyses tools and heavy use of typing have driven me towards these patterns.

If it was purely an optimisation thing, I'd 100% be onboard with what you're saying.